subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java

```diff
@@ -205,11 +212,57 @@ protected void visitLocation(String location) {
 
         @Override
         public void node(Throwable node) {
-            String message = getMessage(node);
-            if (null == node.getCause() || (StringUtils.isNotBlank(message) && !message.endsWith(NO_ERROR_MESSAGE_INDICATOR))) {
-                LinePrefixingStyledTextOutput output = getLinePrefixingStyledTextOutput(failureDetails);
-                renderStyledError(node, output);
+            if (shouldBePrinted(node)) {
+                printedNodes.add(node);
+                if (null == node.getCause() || isUsefulMessage(getMessage(node))) {
+                    LinePrefixingStyledTextOutput output = getLinePrefixingStyledTextOutput(failureDetails);
+                    renderStyledError(node, output);
+                }
+            } else {
+                // Only increment the suppressed branch count for the ultimate cause of the failure, which has no cause itself
+                if (node.getCause() == null) {
+                    suppressedDuplicateBranchCount++;
+                }
+            }
+        }
+
+        /**
+         * Determines if the given node should be printed.
+         *
+         * A node should be printed iff it is not in the {@link #printedNodes} set, and it is not a
+         * transitive cause of a node that is in the set.  Direct causes will be checked, as well
+         * as each branch of {@link ContextAwareException#getReportableCauses()}s for nodes of that type.
+         *
+         * @param node the node to check
+         * @return {@code true} if the node should be printed; {@code false} otherwise
+         */
+        private boolean shouldBePrinted(Throwable node) {
```

Comment:
This seems to be happening at the wrong level because we're looking down into the exception's causes.  Previously, this was just printing the causes as they came in.  Is there some way we can push up this de-duplication so that...
- We get visits for each location
- We get visits for each unique cause
- We get a new type of visit for "there were duplicate causes"

Then the reporter has a simpler job to just render causes and the message when there are duplicates.

Reply:
Are there any unit-level tests that exercise this without having to run the full Gradle build and inspecting the output?

Reply:
Per discussion, I tried summarizing an exception with a loop of causes - it failed with a StackOverflow.  Then I tried rewinding to before my changes and trying it, which failed in the same way, so I don't think this is a problem.  Here's the test case:

```groovy
    def "multi-cause exceptions have branches summarized properly when identical root cause involved in multi-step-loop-cause"() {
        def ultimateCause = new RuntimeException("ultimate cause")
        def intermediate1 = new RuntimeException("intermediate cause 1", ultimateCause)
        def intermediate2 = new RuntimeException("intermediate cause 2", intermediate1)
        ultimateCause.cause = intermediate2
        def branch1 = new DefaultMultiCauseException("first failure", intermediate2)
        def branch2 = new DefaultMultiCauseException("second failure", intermediate2)
        Throwable exception = new ContextAwareException(new DefaultLenientConfiguration.ArtifactResolveException("task dependencies", "org:example:1.0", [branch1, branch2]))

        when:
        reporter.buildFinished(result(exception))
        print(output.value)

        then:
        output.value == """
{failure}FAILURE: {normal}{failure}Build failed with an exception.{normal}

* What went wrong:
Could not resolve all task dependencies for org:example:1.0.
{info}> {normal}first failure
   {info}> {normal}ultimate cause
{info}> {normal}There was 1 additional failure with the same cause that was not printed.

* Try:
$STACKTRACE
$INFO_OR_DEBUG
$SCAN
$GET_HELP
"""
    }
```

Reply:
Pushed the other new tests we talked about - results seem good to me.